dev/com.ibm.ws.microprofile.faulttolerance.2.0/src/com/ibm/ws/microprofile/faulttolerance20/impl/SyncExecutor.java

```diff
@@ -0,0 +1,191 @@
+        if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+            Tr.event(tc, "Fault tolerance execution started for {0}", context.getMethod());
+        }
+
+        SyncExecutionContextImpl executionContext = (SyncExecutionContextImpl) context;
+        Thread runningThread = Thread.currentThread();
+
+        MethodResult<R> result = null;
+        boolean done = false;
+        RetryState retryContext = FaultToleranceStateFactory.INSTANCE.createRetryState(retryPolicy);
+
+        retryContext.start();
+
+        while (!done) {
+            result = null;
+
+            TimeoutState timeoutContext = FaultToleranceStateFactory.INSTANCE.createTimeoutState(executorService, timeoutPolicy);
+
+            if (!circuitBreaker.requestPermissionToExecute()) {
+                if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                    Tr.event(tc, "Method {0} attempt Circuit Breaker open, not executing", context.getMethod());
+                }
+
+                result = MethodResult.failure(new CircuitBreakerOpenException());
+            }
+
+            if (result == null) {
+                timeoutContext.start(() -> runningThread.interrupt());
+
+                result = bulkhead.run(callable);
+
+                if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                    Tr.event(tc, "Method {0} attempt result: {1}", context.getMethod(), result);
+                }
+
+                timeoutContext.stop();
+            }
+
+            if (timeoutContext.isTimedOut()) {
+                if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                    Tr.event(tc, "Method {0} finished but has timed out, result changed to TimeoutException", context.getMethod());
+                }
+
+                result = MethodResult.failure(new TimeoutException());
+                Thread.interrupted(); // Clear interrupted flag if we were timed out
+            }
+
+            circuitBreaker.recordResult(result);
+
+            RetryResult retryResult = retryContext.recordResult(result);
+            if (!retryResult.shouldRetry()) {
+                if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                    Tr.event(tc, "Method {0} not retrying", context.getMethod());
+                }
+
+                done = true;
+            } else {
+                if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                    Tr.event(tc, "Method {0} retrying with delay: {1} {2}", context.getMethod(), retryResult.getDelay(), retryResult.getDelayUnit());
+                }
+
+                try {
+                    Thread.sleep(retryResult.getDelay());
+                } catch (InterruptedException ex) {
+                    done = true; // Stop retrying if we're interrupted
+                }
+            }
+        }
+
+        result = runFallback(result, executionContext);
```

Comment:
comment here since runFallback does not always run a fallback method

Reply:
Done